cmake/Modules/MacroQbtCompilerSettings.cmake

```diff
@@ -1,69 +1,39 @@
+    if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU" OR "${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
+        add_compile_options(
+            -Wall
+            -Wextra
+            -Wcast-qual
+            -Wcast-align
+            -Winvalid-pch
+            -Woverloaded-virtual
+            -Wold-style-cast
+            -Wnon-virtual-dtor
+            -pedantic
+            -pedantic-errors
         )
 
-        include(CheckCXXCompilerFlag)
-        # check for -pedantic
-        check_cxx_compiler_flag(-pedantic _PEDANTIC_IS_SUPPORTED)
-        if (_PEDANTIC_IS_SUPPORTED)
-            list(APPEND _GCC_COMMON_CXX_FLAGS "-pedantic -pedantic-errors")
-        else (_PEDANTIC_IS_SUPPORTED)
-            list(APPEND _GCC_COMMON_CXX_FLAGS "-Wpedantic")
-        endif (_PEDANTIC_IS_SUPPORTED)
-
         if (CMAKE_SYSTEM_NAME MATCHES Linux)
-            add_definitions(-D_DEFAULT_SOURCE)
+            add_compile_definitions(_DEFAULT_SOURCE)
         endif()
 
-        # Clang 5.0 still doesn't support -Wstrict-null-sentinel
+        include(CheckCXXCompilerFlag)
+        # Clang 11 still doesn't support -Wstrict-null-sentinel
         if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
             check_cxx_compiler_flag(-Wstrict-null-sentinel _STRICT_NULL_SENTINEL_IS_SUPPORTED)
             if (_STRICT_NULL_SENTINEL_IS_SUPPORTED)
-                list(APPEND _GCC_COMMON_CXX_FLAGS "-Wstrict-null-sentinel")
-            endif (_STRICT_NULL_SENTINEL_IS_SUPPORTED)
-
-            # Code should be improved to render this not needed
-            list(APPEND _GCC_COMMON_CXX_FLAGS "-Wno-error=unused-function")
-        else ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
+                add_compile_options(-Wstrict-null-sentinel)
+            endif()
+        else()
             # GCC supports it
-            list(APPEND _GCC_COMMON_CXX_FLAGS "-Wstrict-null-sentinel")
-        endif ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
-
-        string(REPLACE ";" " " _GCC_COMMON_C_AND_CXX_FLAGS_STRING "${_GCC_COMMON_C_AND_CXX_FLAGS}")
-        string(REPLACE ";" " " _GCC_COMMON_CXX_FLAGS_STRING "${_GCC_COMMON_CXX_FLAGS}")
-
-        string(APPEND CMAKE_C_FLAGS " ${_GCC_COMMON_C_AND_CXX_FLAGS_STRING}")
-        string(APPEND CMAKE_CXX_FLAGS " ${_GCC_COMMON_C_AND_CXX_FLAGS_STRING} ${_GCC_COMMON_CXX_FLAGS_STRING}")
-    endif ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU" OR "${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang")
-
-    if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
-        set(QBT_ADDITIONAL_FLAGS "/wd4251 /wd4275 /wd4290 /W4" CACHE STRING "Additional qBittorent compile flags")
-    endif ()
-
-    string(APPEND CMAKE_C_FLAGS " ${QBT_ADDITIONAL_FLAGS}")
-    string(APPEND CMAKE_CXX_FLAGS " ${QBT_ADDITIONAL_FLAGS}")
+            add_compile_options(-Wstrict-null-sentinel)
+        endif()
 
-# endif (NOT QBT_ADDITIONAL_FLAGS)
+    elseif (MSVC)
+        add_compile_options(/wd4251 /wd4275 /wd4290 /W4 /MP)
```

Comment:
you really don't need `/MP` here, this will make nmake very slow and maybe make vcproj generation bit weird, I never tried looking for this particular flag in build files but all cmake generators supporting parallel compilation works out of the box without any problem

Reply:
@jagannatharjun 

> parallel compilation works out of the box without any problem

If you mean that CMake sets the "flag that enables parallel jobs int the build tool" to the optimum value automatically based on the the number of logical processors of the machine, that's correct. However, for VS generators that refers to the `/m` flag of MSBuild, which just means multiple projects will be built at the same time, not multiple compilation units. `/MP` must be passed to the compiler directly, not the build tool, for multiple compilation units to be compiled simultaneously.

See the relevant section: https://docs.microsoft.com/en-us/cpp/build/reference/mp-build-with-multiple-processes?view=vs-2019#visual-studio-project-settings

This is corroborated by my own benchmarking (these durations are only meaningful for relative comparison):

- VS 2019 generator with `/MP` forced: ~3:35 minutes
- NMake JOM generator: ~3:10 minutes (note how it's still a touch faster than the VS generator)
- VS 2019 generator (no matter the value set in `--parallel` in the cmake build command line/): 8+ minutes

When using `/MP` vs not using it, I can see the difference in duration, in the CPU usage, in the number of simultaneously spawned compiler processes, and even the sound the PC fan makes.

I encourage you to try it out for yourself, maybe I'm doing something wrong. Here are the commands I use (in a Windows 10 VM with the latest VS2019 toolchain):

```
cmake -S . -B build -G "Visual Studio 16 2019" -A x64 -DCMAKE_TOOLCHAIN_FILE=C:\vcpkg\scripts\buildsystems\vcpkg.cmake -DVCPKG_TARGET_TRIPLET=x64-windows-static
cmake --build .\build\ --config RelWithDebInfo # if no config specified, defaults to Debug
```

Reply:
>  the /MP compiler option can build multiple compilation units at the same time

it will not affect any other "normal" generators like jom, nmake or ninja, since they do individually call compiler for each unit

as far as VS generator is concerned one can add this through the visual studio's option box, let the user handle the choice.

Reply:
@jagannatharjun 

> Docs says
> 
> > the /MP compiler option can build multiple compilation units at the same time
> 
> it will not affect any other "normal" generators like jom, nmake or ninja, since they do individually call compiler for each unit
> 
> as far as VS generator is concerned one can add this through the visual studio's option box, let the user handle it.

I'd rather not rely on the VS GUI - right now it's possible to fully build the project with VS generators without using the GUI at all. I guess I'll add an exception for VS generators. The advantage of VS generators on Windows is that they seem to "just work" with any shell, whereas ninja/nmake/jom only work if I specifically use the "developer command prompt" or otherwise add all relevant paths (for `cl.exe`, for example) manually to `PATH`.

Reply:
Ok, now `/MP` only gets appended to the compile options when using VS generators.